### PR TITLE
v1.0.0 - Migrating to Websockets for real time updates

### PR DESCRIPTION
**Pull request by Xorso**
* Initial changes for moving to websockets
* Updated the way that we store light properties by getting all variables as attributes (will be used for future entities)
* Changes made to Home-Assistant core to support register device callbacks (https://github.com/Xorso/core/tree/control4_wss)

### CODE REVIEW

**Comment by Xorso** [comment]
I am open to design changes. I figured this would just get us started. I working on the sensors proxy now as well so we can get sensors. I don't have any door locks so I am not sure I could implement that but I do have the audio matrix and thermostats that I want to get integrated as well.

**Comment by lawtancool** [comment]
Hi @Xorso! Thanks for your PR!

I'm going to look over the code and test it on my system soon, but for now could you try formatting your code with https://pypi.org/project/black/? This helps keep code in the repo formatted consistently and prevents random formatting changes to the existing code. I'm pretty sure Home Assistant also requires Black formatting for when we make PRs to the core as well.

**Comment by Xorso** [comment]
Interesting. I will update the PR. I did run black before I submitted the PR. I will try again.

File: pyControl4/director.py [hunk 1]
```diff
@@ -50,15 +177,22 @@ async def sendGetRequest(self, uri):
                 async with self.session.get(
                     self.base_url + uri, headers=self.headers
                 ) as resp:
-                    await checkResponseForError(await resp.text())
+                    try:
+                        await checkResponseForError(await resp.text())
+                    except KeyError as err:
```
Comment (lawtancool) [inline pyControl4/director.py:182]: Why was this `KeyError` except added? In what cases does a `KeyError` occur when running `checkResponseForError()`? 

I would prefer to add logic to handle the error properly in the `checkResponseForError()` function instead of just throwing it out to the logger.

**Review by lawtancool — CHANGES_REQUESTED** [review]

File: pyControl4/director.py [hunk 2]
```diff
@@ -87,6 +221,23 @@ async def sendPostRequest(self, uri, command, params, async_variable=True):
                     await checkResponseForError(await resp.text())
                     return await resp.text()
 
+    async def getAllItemsByCategory(self, category, query_string=None):
```
Comment (lawtancool) [inline pyControl4/director.py:224]: Nice find on the Control4 API! Do you have any information on how `query_string` might work?

File: pyControl4/director.py [hunk 3]
```diff
@@ -1,11 +1,75 @@
...
 import json
+import socketio
+import logging
 
 from .error_handling import checkResponseForError
 
+_LOGGER = logging.getLogger(__name__)
+
+
+class C4DirectorNamespace(socketio.AsyncClientNamespace):
```
Comment (lawtancool) [inline pyControl4/director.py:15]: I wonder if we should move this class into a separate .py file and import it into `director.py`. The code in this namespace class seems to be pretty unrelated to the rest of the director functions.

**Comment by lawtancool** [comment]
Hi @Xorso!

I've looked over the code and made a few small comments. 

One design change I was considering is to move all the websockets code (including `C4DirectorNamespace` and all the new callback code under `C4Director`) into a separate file called `websocket.py`. 

The callback code that is currently under `C4Director` would become a new class called `C4Websocket`. Then, in the Home Assistant side, we would use a `C4Websocket` object to register the callbacks:

(`core/__init__.py`)
```
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
...

    director = C4Director(
        config[CONF_HOST], director_token_dict[CONF_TOKEN], director_session
    )
    entry_data[CONF_DIRECTOR] = director

    websocket = C4Websocket(director_token_dict[CONF_TOKEN])
    entry_data[CONF_WEBSOCKET] = websocket
...

class Control4Entity(Entity):
...
    async def async_added_to_hass(self):
        """Sync with HASS."""
        await super().async_added_to_hass()
        await self.hass.async_add_executor_job(
            self.entry_data[CONF_WEBSOCKET].add_device_callback,
            self._idx,
            self._update_callback,
        )
        _LOGGER.debug(f"Registering device {self._device_id} for callback")
        return True
```

What do you think of this?

**Comment by lawtancool** [comment]
@Xorso I ended up implementing the changes I suggested in the previous comment, please take a look and let me know what you think!

If we go ahead with this change, we will have to change the code in the Home Assistant integration to use the `C4Websocket` object as well.

File: pyControl4/director.py (hunk 1 above)
Comment (Xorso) [reply pyControl4/director.py:66]: I hit this while implementing another lookup method for categories. We can strip it out for now if you would like. If was just a safe catch if there were a key error.

**Comment by Xorso** [comment]
Looks good to me. I like the idea of putting it in it's own module. You are right it keeps things cleaner. This was just a first attempt hack and Ithink there are still going to be issues to resolve along the way but It has been working out well for me (there does appear to be some frequent disconnects so we may need to adjust the keep-alive on the socket)

**Comment by Xorso** [comment]
@lawtancool Since you fixed and merged your changes is there anything else you would like me to do with this PR?

**Comment by lawtancool** [comment]
> @lawtancool Since you fixed and merged your changes is there anything else you would like me to do with this PR?

I haven't had time to test the code yet, but I think we should be done soon with this. I'll start looking at the Home Assistant code soon so we can try to merge it into the official repo.

**Comment by Xorso** [comment]
Sounds good. Just wanted to make sure you were not waiting on me.

**Comment by jsb5151** [comment]
Thanks for your work guys.  Any chance we’ll get to see Websocket support for HA soon?  Anything I can do to help test this?

I’m experiencing controller freezes and I’m not sure if it’s related to all the HTTP polling requests sent by HA…

**Comment by Xorso** [comment]
Any change here?

**Review by lawtancool — APPROVED** [review]

**Comment by lawtancool** [comment]
@Xorso I'm going to merge this code and release it as v1.0.0b1 to make it easier to test the Home Assistant integration code. Once the Home Assistant integration is done, I'll re-release with any necessary changes as v1.0.0.

Thanks for your work on this!
